src/src/com/microsoft/aad/adal/UnexpectedServerResponseException.java

```diff
@@ -0,0 +1,14 @@
+package com.microsoft.aad.adal;
+
+import java.io.IOException;
+
+/**
+ * Created by imakhal on 2/17/2016.
+ */
+public class UnexpectedServerResponseException extends IOException {
```

Comment:
I propose that this exception be renamed to ResourceAuthenticationChallengeException.  The goal is to make the name of the exception abstract enough that it could potentially be used for multiple causes, but specific enough that the developer can reasonably apply the same remediation in all cases that the exception is thrown.  The name UnexpectedServerResponseException achieves the first goal, but I don't think that it achieves the second goal.

I would also make the ResourceAuthenticationChallengeException extend the AuthenticationException class instead of the IOException class.  I'd like most, if not all, exceptions thrown by ADAL to be ADAL defined exceptions.  This will make it simpler to maintain a consistent way of collecting telemetry on errors when we introduce a more consistent way of handling sub-errors.  For instance, all Authentication exceptions have an ADALError typed property that can  contain very specific information about the exact cause of the exception in an easily collected enum.